src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingStyle.cs

```diff
@@ -0,0 +1,367 @@
+                case Capitalization.FirstUpper:
+                    return words.Take(1).Select(CapitalizeFirstLetter).Concat(words.Skip(1).Select(DecapitalizeFirstLetter));
+                case Capitalization.AllUpper:
+                    return words.Select(w => w.ToUpper());
+                case Capitalization.AllLower:
+                    return words.Select(w => w.ToLower());
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private string CapitalizeFirstLetter(string word)
+        {
+            var chars = word.ToCharArray();
+            return new string(chars.Take(1).Select(c => char.ToUpper(c)).Concat(chars.Skip(1)).ToArray());
+        }
+
+        private string DecapitalizeFirstLetter(string word)
+        {
+            var chars = word.ToCharArray();
+            return new string(chars.Take(1).Select(c => char.ToLower(c)).Concat(chars.Skip(1)).ToArray());
+        }
+
+        public bool IsNameCompliant(string name, out string failureReason)
+        {
+            if (!name.StartsWith(Prefix))
+            {
+                failureReason = string.Format(FeaturesResources.Missing_prefix_colon_0, Prefix);
+                return false;
+            }
+
+            if (!name.EndsWith(Suffix))
+            {
+                failureReason = string.Format(FeaturesResources.Missing_suffix_colon_0, Suffix);
+                return false;
+            }
+
+            if (name.Length <= Prefix.Length + Suffix.Length)
+            {
+                failureReason = null;
+                return true;
+            }
+
+            var spanToCheck = TextSpan.FromBounds(Prefix.Length, name.Length - Suffix.Length);
+            Debug.Assert(spanToCheck.Length > 0);
+
+            switch (CapitalizationScheme)
+            {
+                case Capitalization.PascalCase: return CheckPascalCase(name, spanToCheck, out failureReason);
+                case Capitalization.CamelCase: return CheckCamelCase(name, spanToCheck, out failureReason);
+                case Capitalization.FirstUpper: return CheckFirstUpper(name, spanToCheck, out failureReason);
+                case Capitalization.AllUpper: return CheckAllUpper(name, spanToCheck, out failureReason);
+                case Capitalization.AllLower: return CheckAllLower(name, spanToCheck, out failureReason);
+                default: throw new InvalidOperationException();
+            }
+        }
+
+        private WordSpanEnumerable GetWordSpans(string name, TextSpan nameSpan)
+            => new WordSpanEnumerable(name, nameSpan, WordSeparator);
+
+        private static string Substring(string name, TextSpan wordSpan)
+            => name.Substring(wordSpan.Start, wordSpan.Length);
+
+        private static Func<string, TextSpan, bool> s_firstCharIsLowerCase = (val, span) => char.IsLower(val[span.Start]);
```

Comment:
You are correct that the logic here is busted. But I was preserving the existing logic. My goal was just to improve perf. I didn't want to change any behavior at all.  Behavior changes can come later